### PR TITLE
Add `BlockQueryWaitTime` Nomad Configuration Option

### LINKED ISSUE #754: Static `WaitTime` Query Values + Server-side Timeouts Lead to Spurious "EOF" Errors
### Blocking Queries + WaitTime
There are a few API calls within the nomad-autoscaler codebase that leverage the [blocking queries](https://developer.hashicorp.com/nomad/api-docs#blocking-queries) feature of the Nomad API. Specifically around listing scaling policies, reading scaling policies, and reading jobs' scaling statuses:

```shellsession
➜  ~/src/hashicorp find nomad-autoscaler -name '*.go' -exec grep -nr -C2 WaitTime {} \;
nomad-autoscaler/plugins/builtin/target/nomad/plugin/state.go-169-      q := &api.QueryOptions{
nomad-autoscaler/plugins/builtin/target/nomad/plugin/state.go-170-              Namespace: jsh.namespace,
nomad-autoscaler/plugins/builtin/target/nomad/plugin/state.go:171:              WaitTime:  5 * time.Minute,
nomad-autoscaler/plugins/builtin/target/nomad/plugin/state.go-172-              WaitIndex: 1,
nomad-autoscaler/plugins/builtin/target/nomad/plugin/state.go-173-      }
nomad-autoscaler/policy/nomad/source.go-90-     s.log.Debug("starting policy blocking query watcher")
nomad-autoscaler/policy/nomad/source.go-91-
nomad-autoscaler/policy/nomad/source.go:92:     q := &api.QueryOptions{WaitTime: 5 * time.Minute, WaitIndex: 1}
nomad-autoscaler/policy/nomad/source.go-93-
nomad-autoscaler/policy/nomad/source.go-94-     for {
--
nomad-autoscaler/policy/nomad/source.go-180-    log.Trace("starting policy blocking query watcher")
nomad-autoscaler/policy/nomad/source.go-181-
nomad-autoscaler/policy/nomad/source.go:182:    q := &api.QueryOptions{WaitTime: 5 * time.Minute, WaitIndex: 1}
nomad-autoscaler/policy/nomad/source.go-183-    for {
nomad-autoscaler/policy/nomad/source.go-184-            var (
```

As-is, these `WaitTime` params have a static defined value of five minutes / 300000ms.

### nomad-autoscaler => Nomad API Connectivity

Our deployment of nomad-autoscaler communicates with its associated Nomad clusters via an AWS load balancer that has an "idle timeout" of 60 seconds ([AWS's default](https://docs.aws.amazon.com/elasticloadbalancing/latest/classic/config-idle-timeout.html#config-idle-timeout-console)). 

## Issue

This static five minute wait time, combined with our specific LB idle timeout arrangements, leads to failed requests anytime that there are no changes to the underlying Nomad state within the timeout duration. E.g., no scaling policy modifications or the like within a minute of initiating a [list scaling policy](https://developer.hashicorp.com/nomad/api-docs/scaling-policies#list-scaling-policies) request. 

Within our autoscaler deployment's logs, this issue manifest in entires like the following:

<details><summary>List Policies</summary>

```json
{
	"message": "encountered an error monitoring policy IDs",
	"attributes": {
		"@level": "error",
		"service": "nomad-autoscaler",
		"@module": "policy_manager",
		"source": "stderr",
		"error": "failed to call the Nomad list policies API: Get \"https://<our_nomad>:4646/v1/scaling/policies?index=47368483&namespace=<our_namespace>&wait=300000ms\": EOF",
	}
}

```
</details>

<details><summary>Get Policy</summary>

```json
{
	"message": "encountered an error monitoring policy",
	"attributes": {
		"@level": "error",
		"@timestamp": "2023-11-02T15:14:09.268724Z",
		"policy_id": "83e13b15-0150-4333-2ed8-f9aa2f7b74c4",
		"service": "nomad-autoscaler",
		"@module": "policy_manager.policy_handler",
		"source": "stderr",
		"error": "failed to get policy: Get \"https://<our_nomad>:4646/v1/scaling/policy/83e13b15-0150-4333-2ed8-f9aa2f7b74c4?index=47299148&namespace=<our_namespace>&wait=300000ms\": EOF",
	}
}
```
</details>

<details><summary>Get Job Scale Status</summary>

```json
{
	"message": "failed to get target status",
	"attributes": {
		"@level": "warn",
		"@timestamp": "2023-11-02T11:30:27.287723Z",
		"policy_id": "83e13b15-0150-4333-2ed8-f9aa2f7b74c4",
		"service": "nomad-autoscaler",
		"@module": "policy_manager.policy_handler",
		"source": "stderr",
		"error": "Get \"https://<our_nomad>:4646/v1/job/infra-experiencer/scale?index=47299822&namespace=cloud-infrastructure-api&wait=300000ms\": EOF",
	}
}
{
	"message": "Get \"https://<our_nomad>:4646/v1/job/infra-experiencer/scale?index=47299822&namespace=<our_namespace>&wait=300000ms\": EOF",
	"attributes": {
		"@level": "error",
		"@timestamp": "2023-11-02T11:30:27.287780Z",
		"policy_id": "83e13b15-0150-4333-2ed8-f9aa2f7b74c4",
		"service": "nomad-autoscaler",
		"@module": "policy_manager.policy_handler",
		"source": "stderr",
	}
}

```
</details>

There is the added nuance of the random `wait / 16` "jitter" delay called out in the Nomad API docs that is also relevant. If the LB idle timeout were five minutes rather than 1 minute, we might still see occasional errors as Nomad might have changes that "unblock" a query after 4.9999 minutes of waiting and then take an additional 0-18.75 seconds to return a response 🙃. Which may sound silly to call out, but it did lead to some confusion on my part when my first attempt at updating `WaitTime` param values to 1 minute didn't resolve all our logged autoscaler errors.

## Desired Outcome

Allow the nomad-autoscaler `WaitTime` values to be configurable. Perhaps as some sort of `blocking_query_wait_time_duration` option under the `nomad {}` configuration block?

### PR DESCRIPTION
**Pull request by jeffwecan**
This PR adds a `BlockQueryWaitTime` config file option / `-nomad-block-query-wait-time` CLI flag to be used when folks want to either:

1. have the autoscaler's Nomad API client wait longer than five minutes for blocking query requests to return a response 
1. or have the client wait a shorter duration for such responses (for instance, if there is a LB idle timeout < 5 minutes in between the autoscaler and associated Nomad API)

I.e., Resolves #754

-------

Aside from the test case updates included as part of this PR's changes, a bit of practical testing was performed as well. Specifically with a shorter-than-5m-default wait time, but still longer than our 60 LB timeout, configured for the autoscaler:
 ```hcl
 nomad {
  namespace = "<our_namespace>"
  block_query_wait_time = "2m"
}
```
 
 Combined with our dev deployment's 60 second idle timeout Nomad LB, we hit the precipitating error but now with our non-default `2m` / `120000ms` wait time reflected (yay):
  
  > failed to call the Nomad list policies API: Get "https://<our_nomad>:4646/v1/scaling/policies?index=47390046&namespace=<our_namespace>&wait=120000ms": EOF
  
  
 Subsequently, we set `block_query_wait_time = "56s"` which was seen to prevent such errors with our particular Nomad API connectivity arrangements.
  
 -------

Also going to link the initial PR adding blocking queries here (for general context + so there is a breadcrumb to this PR in that preceding one): https://github.com/hashicorp/nomad-autoscaler/issues/38

### CODE REVIEW

File: agent/config/config.go [hunk 1]
```diff
@@ -155,6 +155,11 @@ type Nomad struct {
 
 	// SkipVerify enables or disables SSL verification.
 	SkipVerify bool `hcl:"skip_verify,optional"`
+
+	// BlockQueryWaitTime controls how long Nomad API requests supporting blocking queries
+	// are held open. Defaults to 5m.
+	BlockQueryWaitTime    time.Duration
```
Comment (jeffwecan) [inline agent/config/config.go:161]: As a general note regarding the option name choice: I waffled between a few options before settling on this. Primarily motivated by this bit of similar config in the Nomad codebase (though related to a Consul API client setting in that case 🤷🏻): https://github.com/hashicorp/nomad/blob/v1.7.0-beta.1/client/config/config.go#L380-L387

(Happy to call it whatever though 😄 )

**Comment by jeffwecan** [comment]
@jrasell: Would it be possible to get this PR reviewed ahead of the next autoscaler release? 🙏🏻

File: agent/config/config.go (hunk 1 above)
Comment (lgfa29) [reply agent/config/config.go:166]: I think the name looks good to me 👍

**Review by lgfa29 — APPROVED** [review]
Thank you for the PR @jeffwecan!

And apologies for the delay in getting this merged. Nomad 1.7 and Autoscaler HA kept us busy for a while. I'm going over the opened PRs and issues to prepare a new release.

**Pull request merged**